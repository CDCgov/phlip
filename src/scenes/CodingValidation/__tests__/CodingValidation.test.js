import React from 'react'
import { shallow } from 'enzyme'
import { CodingValidation } from '../index'

const props = {
  project: { id: 1, name: 'Project Test', projectJurisdictions: [{ id: 11 }, { id: 10 }, { id: 20, name: 'florida' }] },
  page: 'coding',
  isValidation: false,
  question: { id: 1, text: 'question!' },
  currentIndex: 0,
  questionOrder: [1],
  showNextButton: false,
  isSchemeEmpty: false,
  areJurisdictionsEmpty: false,
  user: { id: 11, role: 'Admin' },
  selectedCategory: null,
  schemeError: null,
  updateAnswerError: null,
  answerErrorContent: null,
  saveFlagErrorContent: null,
  getQuestionErrors: null,
  match: { url: '/project/1/code', params: { id: 1, view: 'code' } },
  history: {
    replace: jest.fn()
  },
  actions: {
    getCodingOutlineRequest: jest.fn(),
    setPage: jest.fn(),
    getValidationOutlineRequest: jest.fn()
  },
  classes: {}
}

describe('CodingValidation', () => {
  test('should render Coding component correctly', () => {
    expect(shallow(<CodingValidation {...props} />)).toMatchSnapshot()
  })
  
  describe('setting page name', () => {
    test('should set document title to code if on coding', () => {
      shallow(
        <CodingValidation
          {...props}
          project={{
            name: 'Blep',
            id: 4,
            projectJurisdictions: []
          }}
        />
      )
      expect(document.title).toEqual('PHLIP - Blep - Code')
    })
    
    test('should set document title to validate if on validation', () => {
      shallow(
        <CodingValidation
          {...props}
          isValidation
          project={{
            name: 'Blep',
            id: 4,
            projectJurisdictions: []
          }}
        />
      )
      expect(document.title).toEqual('PHLIP - Blep - Validate')
    })
  })
  
  describe('getting outline', () => {
    test('should get coding outline if page is coding', () => {
      const spy = jest.spyOn(props.actions, 'getCodingOutlineRequest')
      shallow(<CodingValidation {...props} />)
      expect(spy).toHaveBeenCalled()
    })
    
    test('should get validation outline if page is validation', () => {
      const spy = jest.spyOn(props.actions, 'getValidationOutlineRequest')
      shallow(<CodingValidation {...props} isValidation page="validation" />)
      expect(spy).toHaveBeenCalled()
    })
  })
  
  describe('setting page text when jurisdictions or scheme is empty', () => {
    describe('when the page is coding', () => {
      test('should show no scheme started text is there is no scheme', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: true, areJurisdictionsEmpty: false })
        expect(wrapper.state().startedText).toEqual('You must add questions to the coding scheme before coding.')
      })
  
      test('should show no jurisdictions started text is there are no jurisdictions', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: false, areJurisdictionsEmpty: true })
        expect(wrapper.state().startedText).toEqual('You must add jurisdictions to the project before coding.')
      })
      
      test('should show combined no jur / questions when both sheme and jurisdictions are empty', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: true, areJurisdictionsEmpty: true })
        expect(wrapper.state().startedText).toEqual('You must add jurisdictions and questions to the coding scheme before coding.')
      })
      
      test('should show coordinators must add started text if the user\'s role is Coder', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress user={{ role: 'Coder' }} />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: true, areJurisdictionsEmpty: true })
        expect(wrapper.state().startedText).toEqual('The coordinator for this project has not created a coding scheme or added jurisdictions.')
      })
    })
    
    describe('when the page is validation', () => {
      test('should show no scheme started text is there is no scheme', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress isValidation />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: true, areJurisdictionsEmpty: false })
        expect(wrapper.state().startedText).toEqual('This project doesn\'t have a coding scheme.')
      })
  
      test('should show no jurisdictions started text is there are no jurisdictions', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress isValidation />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: false, areJurisdictionsEmpty: true })
        expect(wrapper.state().startedText).toEqual('This project doesn\'t have jurisdictions.')
      })
  
      test('should show combined no jur / questions when both sheme and jurisdictions are empty', () => {
        const wrapper = shallow(<CodingValidation {...props} getRequestInProgress isValidation />)
        wrapper.setProps({ getRequestInProgress: false, isSchemeEmpty: true, areJurisdictionsEmpty: true })
        expect(wrapper.state().startedText).toEqual('This project does not have a coding scheme or jurisdictions.')
      })
    })
  })
  
  describe('setting jurisdiction', () => {
    test('should set it to empty if there are no project jurisdictions', () => {
      const wrapper = shallow(
        <CodingValidation
          {...props}
          project={{
            id: 4,
            name: 'blep',
            projectJurisdictions: []
          }}
        />
      )
      expect(wrapper.state().jurisdiction).toEqual({ id: null })
    })
    
    test('should use route param if one is present', () => {
      const wrapper = shallow(<CodingValidation {...props} match={{ params: { jid: 20 } }} />)
      expect(wrapper.state().jurisdiction).toEqual({ id: 20, name: 'florida' })
    })
    
    test('should use first jurisdiction if no route param is present', () => {
      const wrapper = shallow(<CodingValidation {...props} />)
      expect(wrapper.state().jurisdiction).toEqual({ id: 11 })
    })
  })
  
  describe('changing routes', () => {
    test('should change routes if component updated and has a new question', () => {
      const spy = jest.spyOn(props.history, 'replace')
      const wrapper = shallow(<CodingValidation {...props} />)
      wrapper.setProps({ question: { id: 32 } })
      expect(spy).toHaveBeenCalledWith({ pathname: '/project/1/code/11/32' })
      spy.mockReset()
    })
    
    test('should change routes if component is updated and has a new jurisdiction', () => {
      const spy = jest.spyOn(props.history, 'replace')
      const wrapper = shallow(<CodingValidation {...props} />)
      wrapper.setState({ jurisdiction: { id: 20 } })
      expect(spy).toHaveBeenCalledWith({ pathname: '/project/1/code/20/1' })
      spy.mockReset()
    })
    
    test('should change routes after load and scheme isn\'t empty', () => {
      const spy = jest.spyOn(props.history, 'replace')
      const wrapper = shallow(<CodingValidation {...props} getRequestInProgress />)
      wrapper.setProps({ getRequestInProgress: false })
      expect(spy).toHaveBeenCalledWith({ pathname: '/project/1/code/11/1' })
      spy.mockReset()
    })
    
    test('should not change the route if scheme or jurisdictions are empty', () => {
      const spy = jest.spyOn(props.history, 'replace')
      shallow(<CodingValidation {...props} isSchemeEmpty areJurisdictionsEmpty />)
      expect(spy).not.toHaveBeenCalled()
    })
  })  
})
