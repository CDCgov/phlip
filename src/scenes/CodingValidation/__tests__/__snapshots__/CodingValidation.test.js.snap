// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodingValidation should render Coding component correctly 1`] = `
<FlexGrid
  align="stretch"
  circular={false}
  container={true}
  flex={true}
  justify="stretch"
  padding={0}
  raised={false}
  style={
    Object {
      "display": "flex",
      "flexWrap": "nowrap",
      "height": "100%",
      "overflow": "hidden",
      "width": "100%",
    }
  }
  type="row"
>
  <Alert
    actions={
      Array [
        Object {
          "onClick": [Function],
          "type": "button",
          "value": "Continue",
        },
      ]
    }
    closeButton={Object {}}
    hideClose={false}
    onCloseAlert={[Function]}
    open={false}
    title="Still Saving Changes"
  >
    <WithStyles(Typography)
      style={
        Object {
          "whiteSpace": "pre-wrap",
        }
      }
      variant="body1"
    >
      We haven't finished saving your answer. If you continue, changes might not be saved.
    </WithStyles(Typography)>
  </Alert>
  <ApiErrorAlert
    actions={
      Array [
        Object {
          "onClick": [Function],
          "type": "button",
          "value": "Try Again",
        },
      ]
    }
    content={null}
    onCloseAlert={[Function]}
    open={false}
  />
  <ApiErrorAlert
    actions={Array []}
    content={null}
    onCloseAlert={[Function]}
    open={false}
  />
  <FlexGrid
    align="stretch"
    circular={false}
    container={true}
    flex={true}
    justify="stretch"
    padding={0}
    raised={false}
    style={
      Object {
        "flexWrap": "nowrap",
        "overflowX": "hidden",
        "overflowY": "auto",
        "width": "100%",
      }
    }
    type="column"
  >
    <withRouter(Header)
      currentJurisdiction={
        Object {
          "id": 11,
        }
      }
      empty={false}
      onGoBack={[Function]}
      onJurisdictionChange={[Function]}
      pageTitle="Coding"
      project={
        Object {
          "id": 1,
          "name": "Project Test",
          "projectJurisdictions": Array [
            Object {
              "id": 11,
            },
            Object {
              "id": 10,
            },
            Object {
              "id": 20,
              "name": "florida",
            },
          ],
        }
      }
    />
    <FlexGrid
      align="stretch"
      circular={false}
      container={true}
      flex={true}
      justify="stretch"
      padding={0}
      raised={false}
      style={
        Object {
          "backgroundColor": "#f5f5f5",
        }
      }
      type="row"
    >
      <FlexGrid
        align="stretch"
        circular={false}
        container={true}
        flex={true}
        justify="stretch"
        padding={0}
        raised={false}
        style={
          Object {
            "overflow": "auto",
          }
        }
        type="row"
      >
        <FlexGrid
          align="stretch"
          circular={false}
          container={true}
          flex={true}
          justify="stretch"
          padding={0}
          raised={false}
          style={
            Object {
              "minHeight": 500,
              "overflow": "auto",
              "padding": "1px 15px 20px 3px",
            }
          }
          type="row"
        >
          <Connect(QuestionCard)
            currentIndex={0}
            getNextQuestion={[Function]}
            getPrevQuestion={[Function]}
            onAnswer={[Function]}
            onApplyAll={[Function]}
            onChange={[Function]}
            onChangeCategory={[Function]}
            onChangeTextAnswer={[Function]}
            onClearAnswer={[Function]}
            onOpenFlagConfirmAlert={[Function]}
            onResetAnswer={[Function]}
            onSave={[Function]}
            onSaveFlag={[Function]}
            page="coding"
            showNextButton={false}
            totalLength={1}
          />
          <Resizable
            defaultSize={
              Object {
                "height": "100%",
                "width": "50%",
              }
            }
            enable={
              Object {
                "bottom": false,
                "bottomLeft": false,
                "bottomRight": false,
                "left": true,
                "right": false,
                "top": false,
                "topLeft": false,
                "topRight": false,
              }
            }
            grid={
              Array [
                1,
                1,
              ]
            }
            handleComponent={
              Object {
                "left": [Function],
              }
            }
            handleStyles={
              Object {
                "left": Object {
                  "bottom": "51.25%",
                  "height": "fit-content",
                  "left": 0,
                  "top": "unset",
                  "width": "fit-content",
                },
              }
            }
            lockAspectRatio={false}
            lockAspectRatioExtraHeight={0}
            lockAspectRatioExtraWidth={0}
            minWidth="10%"
            onResize={[Function]}
            onResizeStart={[Function]}
            onResizeStop={[Function]}
            resizeRatio={1}
            scale={1}
            style={
              Object {
                "display": "flex",
              }
            }
          >
            <FlexGrid
              align="stretch"
              circular={false}
              container={false}
              flex={false}
              justify="stretch"
              padding={0}
              raised={false}
              style={
                Object {
                  "maxWidth": 17,
                  "minWidth": 17,
                  "width": 17,
                }
              }
              type="column"
            />
            <Connect(DocumentList)
              page="coding"
              projectId={1}
              questionId={1}
              saveUserAnswer={[Function]}
            />
          </Resizable>
        </FlexGrid>
      </FlexGrid>
    </FlexGrid>
  </FlexGrid>
  <Alert
    actions={
      Array [
        Object {
          "onClick": [Function],
          "type": "button",
          "value": "Clear Flag",
        },
      ]
    }
    closeButton={Object {}}
    hideClose={false}
    onCloseAlert={[Function]}
    open={false}
    title="Confirm Clear Flag"
  >
    <WithStyles(Typography)
      variant="body1"
    >
      Do you want to clear this flag?
    </WithStyles(Typography)>
  </Alert>
  <ApiErrorAlert
    actions={Array []}
    content={null}
    onCloseAlert={[Function]}
    open={false}
  />
</FlexGrid>
`;
